src/quicklist.c

```diff
@@ -475,23 +476,22 @@ REDIS_STATIC int _quicklistNodeAllowInsert(const quicklistNode *node,
     if (unlikely(QL_NODE_IS_PLAIN(node) || isLargeElement(sz)))
         return 0;
 
-    int ziplist_overhead;
-    /* size of previous offset */
-    if (sz < 254)
-        ziplist_overhead = 1;
-    else
-        ziplist_overhead = 5;
+    size_t lp_overhead = sz;
+    /* size of encode offset */
+    if (sz < 64) lp_overhead += 1;
+    else if (sz < 4096) lp_overhead += 2;
+    else lp_overhead += 5;
 
-    /* size of forward offset */
-    if (sz < 64)
-        ziplist_overhead += 1;
-    else if (likely(sz < 16384))
-        ziplist_overhead += 2;
-    else
-        ziplist_overhead += 5;
+    /* size of backlen offset */
+    if (lp_overhead <= 127) lp_overhead += 1;
+    else if (lp_overhead < 16383) lp_overhead += 2;
+    else if (lp_overhead < 2097151) lp_overhead += 3;
+    else if (lp_overhead < 268435455) lp_overhead += 4;
+    else lp_overhead += 5;
```

Comment:
i don't like all that logic here, maybe it belongs in listpack.c
i'm also not certain (yet) it's correct, but maybe we can get away without it anyway?
SIZE_SAFETY_LIMIT is set to quite low (as far as safety is concerned), and we re-evaluate this condition on each item we add, so the worse case, we'll add an excessive item and breach the configured limit by some 10 bytes?

Reply:
I have tried to put this in the listpack, but it looks weird.
It belongs in the listpack, but not exactly in the listpack, because it's just an estimate.
From the comment below this code `/* new_sz overestimates if 'sz' encodes to an integer type */` we can see that Matt designed it to be an estimate.
I'm in favor of using `SIZE_SAFETY_LIMIT`, and maybe set it to the median value.

Reply:
On second thought, I think it would be more appropriate to delete it (it will only exceed the limit by 10 bytes).